File: frontend/src/types/recipe.ts
// Recipe-related type definitions for the Recipe Finder app

export interface Recipe {
  id: string
  title: string
  description: string
  ingredients: string[]
  instructions: string[]
  prep_time: number // in minutes
  cook_time: number // in minutes
  servings: number
  difficulty: 'easy' | 'medium' | 'hard'
  category: string
  tags: string[]
  image_url?: string
  created_at: string
  updated_at: string
}

export interface CreateRecipeRequest {
  title: string
  description?: string
  ingredients: string[]
  instructions: string[]
  prep_time?: number
  cook_time?: number
  servings?: number
  difficulty?: 'easy' | 'medium' | 'hard'
  category?: string
  tags?: string[]
  image_url?: string
}

export interface UpdateRecipeRequest {
  title?: string
  description?: string
  ingredients?: string[]
  instructions?: string[]
  prep_time?: number
  cook_time?: number
  servings?: number
  difficulty?: 'easy' | 'medium' | 'hard'
  category?: string
  tags?: string[]
  image_url?: string
}

export interface SearchRecipeParams {
  q?: string // query string
  category?: string
  difficulty?: 'easy' | 'medium' | 'hard'
  tags?: string[]
  max_prep_time?: number
  max_cook_time?: number
}

export interface RecipeListResponse {
  recipes: Recipe[]
  total: number
}

export interface SearchRecipeResponse {
  recipes: Recipe[]
  total: number
  query?: string
}

export interface ApiResponse<T = unknown> {
  success: boolean
  data?: T
  message?: string
  error?: string
}

// Component prop types
export interface IngredientInputProps {
  ingredients: string[]
  onChange: (ingredients: string[]) => void
  placeholder?: string
  maxIngredients?: number
  className?: string
  required?: boolean
  label?: string
}

export interface InstructionInputProps {
  instructions: string[]
  onChange: (instructions: string[]) => void
  placeholder?: string
  maxInstructions?: number
  className?: string
  required?: boolean
  label?: string
}

export interface RecipeFormData {
  title: string
  description: string
  ingredients: string[]
  instructions: string[]
  prep_time: number
  cook_time: number
  servings: number
  difficulty: 'easy' | 'medium' | 'hard'
  category: string
  tags: string[]
  image_url: string
}

export interface RecipeCardProps {
  recipe: Recipe
  onEdit?: (recipe: Recipe) => void
  onDelete?: (recipeId: string) => void
  showActions?: boolean
  className?: string
}

export interface RecipeFilterProps {
  onFilterChange: (filters: SearchRecipeParams) => void
  initialFilters?: SearchRecipeParams
  categories?: string[]
  className?: string
}

// Form validation types
export interface FormErrors {
  [key: string]: string | undefined
}

export interface ValidationRule {
  required?: boolean
  minLength?: number
  maxLength?: number
  pattern?: RegExp
  custom?: (value: unknown) => string | undefined
}

export interface ValidationRules {
  [key: string]: ValidationRule
}

// API Error types
export interface ApiError {
  message: string
  status: number
  details?: unknown
}

// Loading states
export type LoadingState = 'idle' | 'loading' | 'success' | 'error'

// Recipe categories and difficulties (can be extended)
export const RECIPE_CATEGORIES = [
  'Italian',
  'Mexican',
  'Asian',
  'Mediterranean',
  'American',
  'Indian',
  'French',
  'Thai',
  'Chinese',
  'Japanese',
  'Korean',
  'Greek',
  'Spanish',
  'German',
  'British',
  'Middle Eastern',
  'African',
  'Caribbean',
  'Vegetarian',
  'Vegan',
  'Gluten-Free',
  'Keto',
  'Paleo',
  'Dessert',
  'Appetizer',
  'Main Course',
  'Side Dish',
  'Breakfast',
  'Lunch',
  'Dinner',
  'Snack',
  'Beverage'
] as const

export const RECIPE_DIFFICULTIES = ['easy', 'medium', 'hard'] as const

export const RECIPE_TAGS = [
  'quick',
  'healthy',
  'comfort-food',
  'party',
  'family-friendly',
  'date-night',
  'budget-friendly',
  'meal-prep',
  'one-pot',
  'no-bake',
  'grilled',
  'baked',
  'fried',
  'steamed',
  'raw',
  'spicy',
  'sweet',
  'savory',
  'creamy',
  'crunchy',
  'protein-rich',
  'low-carb',
  'high-fiber',
  'dairy-free',
  'nut-free',
  'soy-free',
  'egg-free'
] as const

export type RecipeCategory = typeof RECIPE_CATEGORIES[number]
export type RecipeDifficulty = typeof RECIPE_DIFFICULTIES[number]
export type RecipeTag = typeof RECIPE_TAGS[number] 

File: backend/models/recipe.go
package models

import "time"

// Recipe represents a recipe in the system
type Recipe struct {
	ID          string    `json:"id" db:"id"`
	Title       string    `json:"title" db:"title"`
	Description string    `json:"description" db:"description"`
	Ingredients []string  `json:"ingredients" db:"ingredients"`
	Instructions []string `json:"instructions" db:"instructions"`
	PrepTime    int       `json:"prep_time" db:"prep_time"` // in minutes
	CookTime    int       `json:"cook_time" db:"cook_time"` // in minutes
	Servings    int       `json:"servings" db:"servings"`
	Difficulty  string    `json:"difficulty" db:"difficulty"` // easy, medium, hard
	Category    string    `json:"category" db:"category"`
	Tags        []string  `json:"tags" db:"tags"`
	ImageURL    string    `json:"image_url" db:"image_url"`
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time `json:"updated_at" db:"updated_at"`
}

// CreateRecipeRequest represents the request payload for creating a recipe
type CreateRecipeRequest struct {
	Title        string   `json:"title" validate:"required"`
	Description  string   `json:"description"`
	Ingredients  []string `json:"ingredients" validate:"required"`
	Instructions []string `json:"instructions" validate:"required"`
	PrepTime     int      `json:"prep_time"`
	CookTime     int      `json:"cook_time"`
	Servings     int      `json:"servings"`
	Difficulty   string   `json:"difficulty"`
	Category     string   `json:"category"`
	Tags         []string `json:"tags"`
	ImageURL     string   `json:"image_url"`
}

// UpdateRecipeRequest represents the request payload for updating a recipe
type UpdateRecipeRequest struct {
	Title        *string   `json:"title,omitempty"`
	Description  *string   `json:"description,omitempty"`
	Ingredients  *[]string `json:"ingredients,omitempty"`
	Instructions *[]string `json:"instructions,omitempty"`
	PrepTime     *int      `json:"prep_time,omitempty"`
	CookTime     *int      `json:"cook_time,omitempty"`
	Servings     *int      `json:"servings,omitempty"`
	Difficulty   *string   `json:"difficulty,omitempty"`
	Category     *string   `json:"category,omitempty"`
	Tags         *[]string `json:"tags,omitempty"`
	ImageURL     *string   `json:"image_url,omitempty"`
}

// SearchRecipeRequest represents the request parameters for searching recipes
type SearchRecipeRequest struct {
	Query      string   `json:"query"`
	Category   string   `json:"category"`
	Difficulty string   `json:"difficulty"`
	Tags       []string `json:"tags"`
	MaxPrepTime int     `json:"max_prep_time"`
	MaxCookTime int     `json:"max_cook_time"`
} 